src/IonParserTextRaw.ts

```diff
@@ -672,58 +672,75 @@ export class ParserTextRaw {
     }
   }
 
-  private _read_timestamp() : void {
-    var ch;
-    this._start = this._in.position();
-    ch = this._read_N_digits( 4 );      // read year
-    if (ch == CH_T) {
-      ch = this._read();  // this is needed as we disallow yyyyThh:mm - TODO really !?
-    }
-    else if (ch == CH_MS) {
-      ch = this._read_N_digits( 2 );    // read month
-      if (ch == CH_MS) {
-        ch = this._read_N_digits( 2 );  // read day
-      }
-      ch = this._read_optional_time(ch);  // no time unless you at least include month - TODO really !?
-    }
-    ch = this._read_optional_time_offset(ch); // we only allow an offset
-    if (IonText.is_numeric_terminator(ch)) {
-      this._unread(ch);
-      this._end = this._in.position();
-      this._value_push( T_TIMESTAMP );
-    }
-    else {
-      this._error( "invalid character after timestamp" );
-    }
-  }
+    private _readTimestamp() : void {
+        this._start = this._in.position();
+        let ch = this._readPastNDigits(4);//reads past year, throws on non digits.
+        if (ch === CH_T) {
+            this._end = this._in.position();
+            this._value_push( T_TIMESTAMP );
+            return;
+        } else if (ch !== CH_MS) {
+            throw new Error("Timestamp year must be followed by '-' or 'T'.");
+        }
 
-  private _read_optional_time(ch: number) : number {
-    if (ch != CH_T) return ch; // no 'T", no time
-    ch = this._read();
-    if (!IonText.is_numeric_terminator(ch)) {
-      // then it has to be, at least hours and minutes
-      ch = this._read_hours_and_minutes(ch);
-      if (ch == CH_CL) {
-        ch = this._read_N_digits( 2 );         // seconds
-        if (ch == CH_DT) {
-          ch = this._read();
-          ch = this._read_required_digits(ch); // fractional seconds (apparently required)
+        ch = this._readPastNDigits(2);//reads past month, throws on non digits.
+        if (ch === CH_T) {
+            this._end = this._in.position();
+            this._value_push( T_TIMESTAMP );
+            return;
+        } else if (ch !== CH_MS) {
+            throw new Error("Timestamp month must be followed by '-' or 'T'.");
         }
-      }
-    }
-    return ch;
-  }
 
-  private _read_optional_time_offset(ch: number) : number {
-    if (ch == CH_MS || ch == CH_PS) {
-      ch = this._read();
-      ch = this._read_hours_and_minutes( ch ); // of the time offset
-    }
-    else if (ch == CH_Z) {
-      ch =  this._read();
+        ch = this._readPastNDigits(2); //reads past day, throws on non digits.
+        if (IonText.isNumericTerminator(ch)) {
+            this._unread(ch);
+            this._end = this._in.position();
+            this._value_push( T_TIMESTAMP );
+            return;
+        } else if (ch !== CH_T) {
+            throw new Error("Timestamp day must be followed by 'T' or ' '.");
```

Comment:
Not just 'T' or ' '. Any numeric stop character.

Reply:
done